core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java

```diff
@@ -475,9 +480,14 @@ public final ReducedQueryPhase reducedQueryPhase(List<? extends AtomicArray.Entr
         }
         QuerySearchResult firstResult = queryResults.get(0).value.queryResult();
         final boolean hasSuggest = firstResult.suggest() != null;
-        final boolean hasAggs = firstResult.hasAggs();
+        final boolean hasAggs = firstResult.hasAggs() && reducedAggs == null;
         final boolean hasProfileResults = firstResult.hasProfileResults();
-        final List<InternalAggregations> aggregationsList = hasAggs ? new ArrayList<>(queryResults.size()) : Collections.emptyList();
+        final List<InternalAggregations> aggregationsList;
+        if (hasAggs) {
+            aggregationsList = new ArrayList<>(queryResults.size());
+        } else {
+            aggregationsList = reducedAggs == null ? Collections.emptyList() : reducedAggs;
+        }
```

Comment:
I think `hasAggs` is a confusing name given how this method was refactored. Maybe we should restructure the logic a bit with comments to explain what each case maps to? eg. something like

```java
if (reducedAggs != null) {
  // we already have results from intermediate reduces and just need to perform the final reduce
  assert firstResult.hasAggs();
} else if (firstResult.hasAggs()) {
  // the number of shards was less than the buffer size so we reduce agg results directly
} else {
  // no aggregations
}
```